src/features/thai/thaiGlyphComposition.js

```diff
@@ -0,0 +1,40 @@
+/**
+ * Apply Thai Glyph Composition feature to tokens
+ */
+
+import { ContextParams } from '../../tokenizer.js';
+import applySubstitution from '../applySubstitution.js';
+
+/**
+  * Update context params
+  * @param {any} tokens a list of tokens
+  * @param {number} index current item index
+  */
+function getContextParams(tokens, index) {
+    const context = tokens.map(token => token.activeState.value);
+    return new ContextParams(context, index || 0);
+}
+
+/**
+  * Apply Thai required glyphs composition substitutions
+  * @param {ContextRange} range a range of tokens
+  */
+function thaiGlyphComposition(range) {
+    const script = 'thai';
+    let tokens = this.tokenizer.getRangeTokens(range);
+    let contextParams = getContextParams(tokens, 0);
+    contextParams.context.forEach((glyphIndex, index) => {
```

Comment:
Foreach can be inefficent with long arrays as it requires a function call for each element in the array, and function calls can be slow if the compiler doesn't optimize them well. For-in or for loops are guaranteed to run fairly quickly however. This can be fixed later.